nixos/modules/services/system/hot-resize.nix

```diff
@@ -0,0 +1,143 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+
+let
+  cfg = config.services.hotResize;
+
+  resize-script = pkgs.writeShellApplication {
+    name = "hot-resize";
+    runtimeInputs = with pkgs; [
+      cloud-utils
+      e2fsprogs
+      parted
+      util-linux
+      btrfs-progs
+      xfsprogs
+      coreutils
+    ];
+
+    text = ''
+      process_device() {
+        local device="$1"
+        local fstype="$2"
+        local mountpoint="$3"
+
+        echo "Processing $device ($fstype) mounted at $mountpoint..."
+
+        REAL_DEVICE=$(readlink -f "$device")
+        DISK=$(lsblk -ndo pkname "$REAL_DEVICE")
+        PART_NUM=$(lsblk -ndo part "$REAL_DEVICE")
+
+        if [ -z "$DISK" ] || [ -z "$PART_NUM" ]; then
+          echo "Failed to determine disk and partition number for $device"
+          return 1
+        fi
+
+        echo "Growing partition /dev/$DISK partition $PART_NUM..."
+        GROWPART_OUTPUT=$(growpart "/dev/$DISK" "$PART_NUM" 2>&1) || {
+          EXIT_CODE=$?
+          if [ $EXIT_CODE -eq 2 ]; then
+            echo "Partition is already at maximum size, continuing with filesystem resize..."
+          else
+            echo "Failed to grow partition: $GROWPART_OUTPUT"
+            return $EXIT_CODE
+          fi
+        }
+
+        echo "Resizing filesystem..."
+        case "$fstype" in
+          ext4)
+            resize2fs "$device"
+            ;;
+          xfs)
+            xfs_growfs "$mountpoint"
+            ;;
+          btrfs)
+            btrfs filesystem resize max "$mountpoint"
+            ;;
+          *)
+            echo "Unsupported filesystem type: $fstype"
+            return 1
+            ;;
+        esac
+
+        echo "Current size:"
+        df -h "$mountpoint"
+        echo "---"
+      }
+
+      process_device "$@"
+    '';
```

Comment:
This is small for now, but in the future, as this module gets bigger, you may want to consider splitting the bash part into a separate file.

Reply:
Yes i can if you want.